### PR TITLE
Cache preferred resources, use in kubectl resource name autocomplete

### LINKED ISSUE #63145: Completion does not work for some resources like apiservice, clusterrole, resourcequota, etc
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:

completion does not support following resources:

~~~
  * apiservice
  * clusterrole
  * customresourcedefinition
  * limitrange
  * podpreset
  * resourcequota
  * role
~~~

**What you expected to happen**:

~~~
$ kubectl get [TAB][TAB]
~~~

completes above resources.

**How to reproduce it (as minimally and precisely as possible)**:

~~~
$ kubectl get [TAB][TAB]
~~~

**Anything else we need to know?**:

- `apiservice`,`resourcequota` and `clusterrole` are used often and stressful without compltion.
- proposal patch: https://github.com/kubernetes/kubernetes/pull/63144

**Environment**:
- Kubernetes version (use `kubectl version`): Build on 50dd9208372084ef6a09f2647a4f9e8f5857cb1e

### PR DESCRIPTION
**Pull request by liggitt**
Fixes #63145
Fixes https://github.com/kubernetes/kubectl/issues/357
Alternative to #61928 

* starts to unify preferred resource logic on top of ServerGroups()/ServerResourcesForGroupVersion() methods
* allows indicating a cached list of resources is acceptable when calling `kubectl api-resources` (default is still to rediscover)
* uses `kubectl api-resources` in bash completion

```sh
$ kubectl get [TAB][TAB]
apiservices.apiregistration.k8s.io                            networkpolicies.extensions
certificatesigningrequests.certificates.k8s.io                networkpolicies.networking.k8s.io
clusterrolebindings.rbac.authorization.k8s.io                 nodes
clusterroles.rbac.authorization.k8s.io                        persistentvolumeclaims
componentstatuses                                             persistentvolumes
configmaps                                                    poddisruptionbudgets.policy
controllerrevisions.apps                                      pods
cronjobs.batch                                                podsecuritypolicies.extensions
customresourcedefinitions.apiextensions.k8s.io                podsecuritypolicies.policy
daemonsets.apps                                               podtemplates
daemonsets.extensions                                         replicasets.apps
deployments.apps                                              replicasets.extensions
deployments.extensions                                        replicationcontrollers
endpoints                                                     resourcequotas
events                                                        rolebindings.rbac.authorization.k8s.io
events.events.k8s.io                                          roles.rbac.authorization.k8s.io
horizontalpodautoscalers.autoscaling                          secrets
ingresses.extensions                                          serviceaccounts
initializerconfigurations.admissionregistration.k8s.io        services
jobs.batch                                                    statefulsets.apps
limitranges                                                   storageclasses.storage.k8s.io
mutatingwebhookconfigurations.admissionregistration.k8s.io    validatingwebhookconfigurations.admissionregistration.k8s.io
namespaces                                                    volumeattachments.storage.k8s.io
```

```release-note
NONE
```

### CODE REVIEW

**Comment by liggitt** [comment]
@kubernetes/sig-cli-pr-reviews @kubernetes/sig-api-machinery-pr-reviews

**Comment by smarterclayton** [comment]
LGTM

I need to look through the bash completions a bit more and try them out though.

File: pkg/kubectl/cmd/cmd.go [hunk 1]
```diff
@@ -109,7 +109,12 @@ __kubectl_parse_get()
 __kubectl_get_resource()
 {
     if [[ ${#nouns[@]} -eq 0 ]]; then
-        return 1
+      local kubectl_out
+      if kubectl_out=$(kubectl api-resources -o name --cached --request-timeout=5s --verbs=get 2>/dev/null); then
```
Comment (superbrothers) [inline pkg/kubectl/cmd/cmd.go]: In order to make it possible to override `--context` flag, you need to add `$(__kubectl_override_flags)` to arguments of `kubectl api-resources`.

```
 if kubectl_out=$(kubectl api-resources $(__kubectl_override_flags) -o name --cached --request-timeout=5s --verbs=get 2>/dev/null); then
```

Comment (liggitt) [reply pkg/kubectl/cmd/cmd.go]: thanks, updated. also, fixed a bug in `__kubectl_override_flags()` that was making use of `--all-namespaces` (which isn't supported for any of the places `__kubectl_override_flags` was being used (config and api-resources don't support that flag, get containers and get resource cannot get by name when combined with --all-namespaces.

**Pull request merged**